File: Server/Serval.Server/Cameras/Camera.cs
using System.Text.Json.Serialization;
using MongoDB.Bson.Serialization.Attributes;

namespace Serval.Server.Cameras;

/// <summary>
/// A camera the server ingests from. Its video comes from one or more <see cref="Streams"/>, each
/// tagged with what it is for — so a camera that offers both a high-quality main stream and a
/// cheap sub stream can be recorded at full quality while detection runs on the small one. The id
/// is also the media subdirectory name and the path segment the CameraModule POSTs telemetry to,
/// so it must be filesystem- and URL-safe.
/// </summary>
[BsonIgnoreExtraElements] // tolerate fields a newer schema may add, rather than crash reads
public sealed class Camera
{
    [BsonId]
    public required string Id { get; set; }

    public required string Name { get; set; }

    public string? Location { get; set; }

    /// <summary>
    /// The camera's sources. Exactly one stream must carry <see cref="StreamRole.Detect"/> and one
    /// <see cref="StreamRole.Live"/>; at most one carries <see cref="StreamRole.Record"/>, and a
    /// camera with none keeps nothing. A stream may carry no roles at all, in which case it is
    /// stored and never pulled. A camera with a single stream therefore declares one entry with
    /// <c>["record","detect","live"]</c>; a camera offering a sub stream typically gives it
    /// <c>["detect"]</c> and keeps <c>["record","live"]</c> on the main.
    /// </summary>
    public required List<CameraStream> Streams { get; set; }

    /// <summary>Disabled cameras are registered but not ingested.</summary>
    public bool Enabled { get; set; } = true;

    /// <summary>
    /// Write this camera's footage to disk. True by default — a camera with a record stream records.
    ///
    /// Separate from <see cref="StreamRole.Record"/> because the two answer different questions.
    /// The role says *which* stream would be written, and survives being switched off, so turning
    /// recording back on does not mean deciding that again — which is what makes this a temporary
    /// switch rather than a reconfiguration. Validation rejects true with no record stream, so it
    /// is never on and inert.
    ///
    /// Footage already on disk is untouched either way: it stays playable and ages out under
    /// <see cref="RetentionDays"/>.
    /// </summary>
    public bool Recording { get; set; } = true;

    /// <summary>Per-camera retention override; null falls back to the server default.</summary>
    public int? RetentionDays { get; set; }

    /// <summary>
    /// ONVIF device service URL (e.g. <c>http://192.168.1.50:80/onvif/device_service</c>). Set it,
    /// with <see cref="OnvifUsername"/>/<see cref="OnvifPassword"/>, to enable PTZ control. The
    /// server discovers the camera's PTZ service and media profile from here. Null means no PTZ.
    /// </summary>
    public string? OnvifUrl { get; set; }

    /// <summary>ONVIF account used for PTZ SOAP calls (WS-Security UsernameToken).</summary>
    public string? OnvifUsername { get; set; }

    /// <summary>Password for <see cref="OnvifUsername"/>. Sent only as a WS-Security digest, never in clear.</summary>
    public string? OnvifPassword { get; set; }

    /// <summary>
    /// Media profile token to drive PTZ against. Optional: when null the server uses the camera's
    /// first PTZ-capable profile, discovered via ONVIF. Set it to pin a specific profile.
    /// </summary>
    public string? OnvifProfileToken { get; set; }

    /// <summary>
    /// Opt in to two-way audio (talk-back) for cameras that support an RTSP/ONVIF backchannel.
    /// When true, go2rtc keeps its backchannel enabled so a viewer's microphone reaches the camera
    /// over the same WebRTC session as the live view. When false the backchannel is disabled at the
    /// source (<c>#backchannel=0</c>) — go2rtc probes it by default, and that probe breaks some
    /// cameras (e.g. certain doorbells), so talk-back is off unless a camera is known to handle it.
    /// </summary>
    public bool TwoWayAudio { get; set; }

    /// <summary>
    /// Record the camera's audio track alongside its video, in the same segment files.
    ///
    /// Off by default and per-camera, not global: recording sound is treated differently from
    /// recording pictures in many jurisdictions, so it is an explicit decision per camera rather
    /// than something that arrives with an upgrade. A camera whose source has no audio track
    /// records video regardless.
    /// </summary>
    public bool RecordAudio { get; set; }

    /// <summary>
    /// Where this camera's volume slider starts, as a lift in dB above its recorded level. 0 starts it
    /// at full volume with nothing added.
    ///
    /// A starting point, not the applied gain. Nothing on the Server reads it, and the App reads it
    /// only until a client has a position of its own: the slider is kept per camera on the machine
    /// doing the listening, because how loud you want to listen is a property of what you are sitting
    /// at and syncing it would let a phone dictate a desktop's volume. This is what stops a camera
    /// somebody has already calibrated from arriving silent on every new browser.
    ///
    /// It lives on the camera because how far a camera needs lifting is a property of its microphone
    /// rather than of who is listening: measured across this deployment, the quietest camera's audio
    /// sits 15-20 dB below the others at every percentile, so one number shared by all of them is
    /// wrong for at least one.
    ///
    /// Applied at playback rather than baked in at record time so it works on footage already on
    /// disk, costs nothing, and stays reversible. Recording it in would also be irreversible for the
    /// rare loud transient: every camera here reaches within a few dB of full scale occasionally, and
    /// a permanent gain would clip those forever.
    /// </summary>
    public double PlaybackGainDb { get; set; }

    /// <summary>
    /// The level below which this camera's audio is treated as silence during playback; null gates
    /// nothing.
    ///
    /// The companion to <see cref="PlaybackGainDb"/> and the reason a large gain is usable at all.
    /// These streams spend most of their time on the noise floor, so a gain big enough to make the
    /// quiet content audible also amplifies the codec's own quantisation noise into audible hiss. A
    /// gate ahead of the gain keeps silence silent. There is room for it: the floor and the content
    /// are separated by 23 dB or more on every camera here.
    ///
    /// Note a compressor would be the wrong instrument, not a gentler one — an AGC applies its
    /// greatest gain exactly when there is nothing but noise to find.
    ///
    /// Expressed as the RMS of a 16 kHz window, the same unit as
    /// <see cref="CameraAudioTuning.SpeechGateRmsThreshold"/>, so it reads against the same live
    /// meter. It is otherwise unrelated to those: this one feeds the speakers, they feed the detector.
    /// </summary>
    [BsonIgnoreIfNull]
    public double? PlaybackGateRmsThreshold { get; set; }

    /// <summary>
    /// Run server-side scene description for this camera, gated on motion.
    ///
    /// This is the point of the shared detection library: a camera with no edge module still gets
    /// AI, run by the Server on its behalf. Off by default because the vision model costs seconds
    /// of CPU per description and one model instance is shared across every camera.
    /// </summary>
    public bool AiVision { get; set; }

    /// <summary>
    /// Run server-side transcription, speaker labelling and conversation reprocessing for this
    /// camera. Requires the camera to actually carry an audio track; the Server pulls it in a
    /// second, audio-only RTSP session so a fault there cannot disturb recording.
    /// </summary>
    public bool AiAudio { get; set; }

    /// <summary>
    /// Per-camera overrides for the audio detection thresholds; null falls back to the server
    /// defaults under <c>Serval:Ai</c>. See <see cref="CameraAudioTuning"/> for why the right
    /// value is a property of the room rather than something the server can pick once.
    /// </summary>
    [BsonIgnoreIfNull]
    public CameraAudioTuning? AudioTuning { get; set; }

    /// <summary>
    /// Per-camera overrides for object detection; null falls back to the server defaults under
    /// <c>Serval:Ai:Detection</c>. See <see cref="CameraDetectionTuning"/> — masks especially have
    /// no sensible global value, because where a property line runs is a fact about one camera.
    /// </summary>
    [BsonIgnoreIfNull]
    public CameraDetectionTuning? DetectionTuning { get; set; }

    /// <summary>
    /// Per-camera overrides for which sounds this camera reports and which of them are alarming;
    /// null falls back to <c>Serval:Ai:Sound</c>. See <see cref="CameraSoundTuning"/> — a driveway
    /// and a nursery want different alert labels, and one list serves both badly.
    /// </summary>
    [BsonIgnoreIfNull]
    public CameraSoundTuning? SoundTuning { get; set; }

    /// <summary>
    /// Per-camera overrides for the movement gate; null falls back to <c>Serval:Ai:Motion</c>. Only
    /// reached on a server with object detection switched off, which is where it is the only thing
    /// deciding whether the description model runs — see <see cref="CameraMotionTuning"/>.
    /// </summary>
    [BsonIgnoreIfNull]
    public CameraMotionTuning? MotionTuning { get; set; }

    /// <summary>
    /// Whether this camera is *set up* for PTZ — an ONVIF endpoint is configured, so there is
    /// something to ask.
    ///
    /// Deliberately not named for capability: it says nothing about which axes the camera has, or
    /// whether it has any at all. A fixed-lens pan/tilt dome and a motorised zoom both answer true
    /// here. For what the camera can actually do, probe
    /// <c>GET /api/cameras/{id}/ptz/capabilities</c>, which asks the camera itself.
    /// </summary>
    [BsonIgnore]
    public bool PtzConfigured => !string.IsNullOrWhiteSpace(OnvifUrl);

    /// <summary>
    /// A copy with the camera's own credentials removed, for callers who are allowed to see that a
    /// camera exists but not to log in to it.
    ///
    /// There are two of them, and the second is the one that gets forgotten: the ONVIF password
    /// sits in its own field, and the stream URLs carry another in their userinfo
    /// (<c>rtsp://admin:hunter2@…</c>). Both are usually the camera's *administrative* credentials
    /// and are usually reused across every camera in the house, so handing them to an account that
    /// exists only to watch the wall gives that account the cameras themselves — outside Serval,
    /// where none of these roles apply.
    ///
    /// A copy rather than an edit in place: the instances this is called on come straight from the
    /// registry, and a redaction that leaked back into a write would erase the real credential.
    /// </summary>
    public Camera WithoutSecrets()
    {
        var copy = (Camera)MemberwiseClone();
        copy.OnvifPassword = null;
        copy.Streams = [.. Streams.Select(stream => stream.WithoutSecrets())];
        return copy;
    }

    /// <summary>
    /// The stream written to disk and played back, or null for a camera that keeps nothing.
    /// </summary>
    [BsonIgnore]
    [JsonIgnore]
    public CameraStream? RecordStream => Pick(StreamRole.Record);

    /// <summary>
    /// Where snapshots, motion detection, the dashboard wall and the AI's audio come from.
    /// </summary>
    [BsonIgnore]
    [JsonIgnore]
    public CameraStream? DetectStream => Pick(StreamRole.Detect);

    /// <summary>What go2rtc serves over WebRTC for the focused single-camera view.</summary>
    [BsonIgnore]
    [JsonIgnore]
    public CameraStream? LiveStream => Pick(StreamRole.Live);

    /// <summary>
    /// The stream that declared this role, or null.
    ///
    /// For <see cref="StreamRole.Record"/> null is an ordinary camera that keeps nothing. For the
    /// other two it is reachable only for a camera the API never accepted — a document written
    /// straight into Mongo. There is deliberately no fallback to the recorded stream: that would be
    /// invisible over the API, so a camera could be decoding its 4K main stream once a second for
    /// thumbnails, or serving it over WebRTC, with nothing anywhere saying so. Every consumer
    /// null-guards, and <see cref="CameraRegistryCheck"/> is what makes such a camera visible
    /// instead of silent.
    /// </summary>
    private CameraStream? Pick(StreamRole role) =>
        Streams.FirstOrDefault(s => s.Roles.Contains(role));
}

File: App/serval_app/lib/models/camera.dart
import 'package:flutter/painting.dart';

/// What is known about reaching a camera right now.
///
/// Three states rather than a bool, and [connecting] is the one that earns the
/// enum: the Server publishes no status field, so everything here is derived
/// from whether that camera's frames are still arriving — and silence has two
/// completely different causes. A camera that has stopped sending is down. A
/// wall that has just started listening has heard nothing from *any* camera
/// yet, and saying six cameras are down on the strength of one socket that has
/// not finished opening is a claim nothing supports.
///
/// So [offline] is a failure state and nothing weaker: we have been listening
/// long enough that silence is the camera's own.
enum CameraConnection {
  /// Frames are arriving.
  online,

  /// Nothing has arrived yet, and not enough time has passed to call that a
  /// fault — a cold start, a reconnect, or the App coming back from the
  /// background.
  connecting,

  /// We have been listening and heard nothing. This one really is down.
  offline,
}

/// A camera, shaped after the Server's `Camera` entity so the wiring pass is a
/// swap rather than a rewrite.
///
/// The Server serializes that entity in camelCase — `id`, `twoWayAudio`,
/// `ptzConfigured` and so on. Telemetry documents use snake_case instead, so
/// resist giving these models one shared naming strategy.
class Camera {
  const Camera({
    required this.id,
    required this.name,
    this.enabled = true,
    this.twoWayAudio = false,
    this.recordAudio = false,
    this.playbackGainDb = 0,
    this.playbackGateRms,
    this.aiVision = false,
    this.aiAudio = false,
    this.ptzConfigured = false,
    this.connection = CameraConnection.connecting,
    this.isRecording = false,
    this.records = true,
    this.hasAudioActivity = false,
    this.needsAttention = false,
    this.resolutionLabel,
    required this.placeholder,
  });

  /// Client-supplied on the Server, constrained to `[A-Za-z0-9_-]+` because it
  /// doubles as a directory name and a URL segment.
  final String id;
  final String name;

  /// Registered but not ingested when false.
  final bool enabled;

  /// The talk-back opt-in. go2rtc only probes the camera's audio backchannel
  /// when this is set, so it gates the *Hold to talk* control.
  final bool twoWayAudio;
  final bool recordAudio;

  /// How far to lift this camera's audio when listening to it, in dB. 0 plays it
  /// as recorded.
  ///
  /// On the camera rather than on this machine because it describes the
  /// microphone, not the listener: cameras on one server can differ by 20 dB or
  /// more in how loudly they record the same room, so one number for all of them
  /// is wrong for at least one. The listening level — the volume slider's 0..1 —
  /// stays local, and the two multiply.
  final double playbackGainDb;

  /// The level below which this camera's audio is silence during playback, as the
  /// RMS of a 16 kHz window; null gates nothing.
  ///
  /// The companion to [playbackGainDb], and the reason a large gain is usable: a
  /// camera stream spends most of its life on its noise floor, so a gain big
  /// enough to make the quiet content audible also lifts the codec's own
  /// quantisation noise into hiss. Gating ahead of the gain keeps silence silent.
  ///
  /// Unrelated to the identically-shaped thresholds under `audioTuning` despite
  /// sharing their unit: this one feeds the speakers, those feed the detector.
  final double? playbackGateRms;

  final bool aiVision;
  final bool aiAudio;

  /// Computed Server-side from the presence of an ONVIF endpoint — so it says
  /// only that there is something to *ask*, not what the camera can do.
  ///
  /// It does not gate the pad: a fixed bullet camera with an ONVIF endpoint
  /// answers true here and has no pan, tilt or zoom at all. What the controls
  /// are drawn from is `ServalRepository.ptzProbeFor`, which asks the camera.
  final bool ptzConfigured;

  /// Derived from snapshot staleness on `WS /api/dashboard` rather than from a
  /// status field, because the Server exposes none — see [CameraConnection] for
  /// why silence needs three answers rather than two.
  ///
  /// [CameraConnection.online] is also the health signal: a frame arriving says
  /// this camera's pipeline is running, and says nothing about what that
  /// pipeline does with the frames.
  ///
  /// Defaults to [CameraConnection.connecting], which is what a camera nobody
  /// has heard from yet actually is.
  final CameraConnection connection;

  /// Exact rather than inferred, even though no per-camera recording state is
  /// published: one ffmpeg per camera produces the snapshots *and* the segments
  /// from a single connection, so a fresh frame is evidence the writer is up.
  ///
  /// [CameraConnection.online] and [records] together — the pipeline is alive
  /// *and* it was asked to write. False whenever [records] is: a camera set to
  /// keep nothing is never recording, however fresh its frames are.
  final bool isRecording;

  /// Whether this camera is *configured* to record, from its `record` stream
  /// role.
  ///
  /// Distinct from [isRecording], and the distinction is the point: false here
  /// means recording was never asked for, false there can also mean a recorder
  /// that has stopped. "Nothing is kept" and "the recorder fell over" want
  /// opposite reactions from whoever is reading the screen.
  final bool records;

  /// Derived from live `utterance` telemetry on `WS /api/events` — a camera
  /// that has heard speech in the last few seconds shows the waveform mark.
  final bool hasAudioActivity;

  /// An alert **sound** on `WS /api/events` in the last few seconds.
  ///
  /// Sounds only, which is narrower than the alerts the rest of the UI counts:
  /// a detection carries `is_alert` too and lights the activity card and the
  /// scrubber, but not the tile it happened on.
  final bool needsAttention;

  /// STUB: the camera model carries no resolution, and the snapshot cannot
  /// stand in — `Ingest:SnapshotMaxMegapixels` fits it to a pixel budget, so a
  /// decoded frame reports the thumbnail's size. It would have to join the DTO.
  final String? resolutionLabel;

  /// How this camera's tile paints until a real frame arrives.
  final TilePlaceholder placeholder;

  /// The centered label the design paints over a tile with no frame yet —
  /// `DRIVEWAY · 1080p`, `BACK YARD · NIGHT`.
  String get placeholderLabel {
    final upper = name.toUpperCase();
    return resolutionLabel == null ? upper : '$upper · $resolutionLabel';
  }
}

/// The diagonal-stripe fill a tile shows before its first frame.
///
/// The design gives every camera its own two-tone stripe and a colored bloom,
/// so a wall of placeholder tiles still reads as six distinct cameras. The
/// wiring pass replaces only the tile body, so this survives as the loading
/// and reconnecting state.
class TilePlaceholder {
  const TilePlaceholder({
    required this.stripeLight,
    required this.stripeDark,
    required this.bloom,
  });

  /// A stable placeholder for a real camera, picked from the design's own five.
  ///
  /// The design hand-picks a stripe and bloom per camera; a registry read gets a list of ids
  /// instead. Hashing the id preserves what the choice was actually for — that a wall of tiles
  /// with no frames yet still reads as several distinct cameras — while staying stable across
  /// runs, so a camera does not change colour when the Server reorders its list.
  factory TilePlaceholder.forCameraId(String cameraId) {
    // FNV-1a rather than String.hashCode: Dart's is seeded per isolate, so it would give a
    // camera a different colour on every launch.
    var hash = 0x811c9dc5;
    for (final unit in cameraId.codeUnits) {
      hash = ((hash ^ unit) * 0x01000193) & 0xFFFFFFFF;
    }
    return _palette[hash % _palette.length];
  }

  static const _palette = [
    TilePlaceholder(
      stripeLight: Color(0xFF1E2131),
      stripeDark: Color(0xFF181B28),
      bloom: Color(0x129184D9),
    ),
    TilePlaceholder(
      stripeLight: Color(0xFF22212C),
      stripeDark: Color(0xFF1B1A24),
      bloom: Color(0x17E0955F),
    ),
    TilePlaceholder(
      stripeLight: Color(0xFF1A2028),
      stripeDark: Color(0xFF161B22),
      bloom: Color(0x0FB4DCD2),
    ),
    TilePlaceholder(
      stripeLight: Color(0xFF20222E),
      stripeDark: Color(0xFF1A1C25),
      bloom: Color(0x0DE9E9ED),
    ),
    TilePlaceholder(
      stripeLight: Color(0xFF1E2131),
      stripeDark: Color(0xFF181B28),
      bloom: Color(0x0D9184D9),
    ),
  ];

  /// What a camera that has dropped out paints: flat, no stripes, no bloom, so an absent feed
  /// never reads as a dark one.
  static const offline = TilePlaceholder(
    stripeLight: Color(0xFF14161F),
    stripeDark: Color(0xFF14161F),
    bloom: Color(0x00000000),
  );

  final Color stripeLight;
  final Color stripeDark;

  /// The radial bloom laid over the stripes, centered a little above middle.
  final Color bloom;
}

/// Where a tile sits on the wall's grid, and how much of it the tile keeps.
///
/// The design's rule is that tiles keep their own size and shape — a driveway
/// worth watching stays 2x2 while a garage stays 1x1 — so position and span
/// are data, not layout code.
class TileLayout {
  const TileLayout({
    required this.cameraId,
    required this.column,
    required this.row,
    this.columnSpan = 1,
    this.rowSpan = 1,
  });

  final String cameraId;

  /// Zero-based grid position.
  final int column;
  final int row;
  final int columnSpan;
  final int rowSpan;

  TileLayout copyWith({int? column, int? row, int? columnSpan, int? rowSpan}) =>
      TileLayout(
        cameraId: cameraId,
        column: column ?? this.column,
        row: row ?? this.row,
        columnSpan: columnSpan ?? this.columnSpan,
        rowSpan: rowSpan ?? this.rowSpan,
      );

  /// By value, so `listEquals` can answer whether a drag actually changed
  /// anything and so a failing layout test prints the two arrangements rather
  /// than two instance hashes.
  @override
  bool operator ==(Object other) =>
      other is TileLayout &&
      other.cameraId == cameraId &&
      other.column == column &&
      other.row == row &&
      other.columnSpan == columnSpan &&
      other.rowSpan == rowSpan;

  @override
  int get hashCode => Object.hash(cameraId, column, row, columnSpan, rowSpan);

  @override
  String toString() =>
      'TileLayout($cameraId at $column,$row size ${columnSpan}x$rowSpan)';
}

/// A detection drawn over the video, as fractions of the frame.
///
/// Built from a `DetectionDocument`'s geometry — see `DetectionDocument.overlays`.
/// One episode yields as many of these as there were objects of its class, since
/// three people in shot is one episode and three boxes. Scene descriptions still
/// carry no geometry, and never will: they are prose about a moment, not a
/// measurement of one.
class DetectionBox {
  const DetectionBox({
    required this.label,
    required this.confidence,
    required this.rect,
    this.isAlert = true,
    this.isStale = false,
  });

  final String label;

  /// Whether the episode this came from was a claim that someone should look.
  ///
  /// Carried on the box because it is what picks its colour: an alert is drawn in `Serval.alert`,
  /// anything else in the Nocturne accent. Only a screen asking for the wide overlay produces the
  /// second kind, so it defaults to true — every box that reaches the screen without one being
  /// asked for is an alert.
  final bool isAlert;

  /// Whether this is the last position known rather than one measured at the instant on screen.
  ///
  /// True inside a gap: the episode covers this moment and the object was not seen in it. Drawn
  /// as a dotted outline instead of a solid one, which is what lets the overlay show where
  /// something was without claiming to have just seen it there — the alternative was a box that
  /// blinked out while the activity row still read "still there".
  ///
  /// Independent of [isAlert]. A dotted box keeps its colour, because how sure we are of a
  /// position and whether the thing in it matters are different questions.
  final bool isStale;

  /// How well *this* object was seen, not how well the episode's best was. Three
  /// people in one episode are three boxes with three scores; writing the
  /// episode's peak over each would be a measurement of one of them presented as
  /// a measurement of all three.
  final double confidence;

  /// Left/top/width/height as fractions of the frame, so the box tracks the
  /// video however it is scaled.
  final Rect rect;

  /// The label, and how well it was seen — but only while that is a reading of the frame you are
  /// looking at. A stale box's confidence was measured somewhere else in time and would argue
  /// against the dotted outline drawn around it, so it says the label alone.
  String get caption => isStale
      ? label.toUpperCase()
      : '${label.toUpperCase()} · ${confidence.toStringAsFixed(2)}';
}
